### ISSUE TITLE
Rename METADATA.jl to CuratedPackages.jl

### ISSUE THREAD

**Issue by milktrader** [JuliaLang/METADATA.jl#2800]
This will make the curated repository of repositories more descriptive, thereby complying with its own naming guidelines :smile:

**Comment by IainNZ** [comment]
Well, assuming we want it to be a curated, which is still ?!. I think a curated repository actually should be a new one, because otherwise we'd have to throw things out of this one.

**Comment by StefanKarpinski** [comment]
Totally possible, but we need to come up with a coherent strategy and not just rename the repo.

**Comment by milktrader** [comment]
Yep, just changing the name would a bit of a mess. I like the idea of simply creating a new repo of repos though and beginning to migrate packages to it. Also, the new package could be explicit in what it is, and the guidelines for inclusion. This repo could begin enforcing minimum requirements (License, REQUIRE with a Julia version, passing tests (?), reasonable package name, etc) right from the start.

So maybe decide on these requirements, migrate packages from METADATA that meet them, and loosen up inclusion rules for METADATA.

**Comment by tkelman** [comment]
Then a question is do we make CuratedPackages the default, and free-for-all METADATA opt-in? The free-for-all could be called something like OpenRegistry.jl if we also feel like renaming it too.

**Comment by milktrader** [comment]
I like OpenRegistry as a name. Yes, I think CuratedPackages should be the default. OpenRegistry could include a dozen implementations of gradient descent, many of which are students just doing their homework assignments in the form of a package. CuratedPackages would have the best-of-breed package.

**Comment by milktrader** [comment]
This requires a small change to the `Pkg.add()` method to include a kwarg for `repo="CuratedPackages"`.

**Comment by milktrader** [comment]
https://github.com/JuliaLang/julia/issues/11914

**Comment by quinnj** [comment]
-1. This sounds like potentially a *lot* of work + resulting breakage/issues for not a lot of gain. I feel like I'm really against having an "OpenRegistry" repo. That's something that someone could start curating on their own if they really wanted without the need to provide something "official" (i.e. Julia.jl). I think it'd be much better to, for example, make `Pkg.clone()` resolve declared dependencies; then `Pkg.clone()` becomes my true free-for-all that acts almost exactly like `Pkg.add` while `Pkg.add` can continue to pull from METADATA where things are more carefully curated (particularly with #2777)

**Comment by ssfrr** [comment]
I'm also a little unsure about the value of a free-for-all package registry. `Pkg.clone()` is so easy to do, and maybe if we don't want to resolve dependencies by default there could be a `recursive=true` option to turn it on.

**Comment by dpsanders** [comment]
I agree that the idea of a registry for students' homework, in @milktrader's example, does not sound appealing. 

How about `dependencies=true` instead of `recursive=true`?

**Comment by milktrader** [comment]
Why not eliminate `Pkg.clone()` and simply allow users to call a `repo=` from `Pkg.add()`. 

R's `install.packages()` includes a repo argument, amongst other things. I don't recall if it defaults to CRAN, but regardless, `Pkg.add()` could default to the curated Julia repo of repos.

**Comment by milktrader** [comment]
@quinnj I agree we don't need to make more work for ourselves, but perhaps an OpenRegistry.jl could be automated to allow just about anything within reason.

It might also be better for someone _outside_ of JuliaLang to create this repo of repos if there is a need for it.

**Comment by quinnj** [comment]
Good thought @milktrader. I wonder if you could just do `Pkg.add("quinnj/CSV.jl")` which would expand to the current equivalent of `Pkg.clone("https://github.com/quinnj/CSV.jl")`. Or perhaps we have `Pkg.add("quinnj","CSV.jl")` with the repo as the 1st argument. In terms of handling dependencies of non-METADATA.jl packages, however, I think we need a mechanism to resolve dependencies of other non-METADATA packages. I.e. in my REQUIRES file, how can I declare a dependency on another non-registered package.

**Comment by milktrader** [comment]
I'd like to keep the current dispatch and just add a kwarg,
````julia
Pkg.add(pkg::AbstractString, vers::VersionNumber... ;
        repos="https://github.com/JuliaLang/METADATA.jl")
````
And I suppose it's not a big deal to pass "CuratedPackages" to the `repos` arg and have it resolve to the actual url.

**Comment by tkelman** [comment]
> I think we need a mechanism to resolve dependencies of other non-METADATA packages. I.e. in my REQUIRES file, how can I declare a dependency on another non-registered package.

I think we do need to support unregistered url's in REQUIRE, and might be nice to allow the version specifier to take a gitsha as well both in REQUIRE and `Pkg.add`. Though we'd need to be very careful to filter out any packages with (non-curated) unregistered dependencies from the curated version.

**Comment by milktrader** [comment]
About resolving dependancies, I'm not sure the time is now, but eventually Julia will need an independent package management system similar to Ruby's `rvm` and Python's `virtualenv`.

**Comment by tkelman** [comment]
https://github.com/rened/DeclarativePackages.jl has some good ideas here.

**Comment by milktrader** [comment]
ping @rened

**Comment by sjkelly** [comment]
maybe we could do something like:
```
using <Pkg> <SHA>
```
Which would be more akin to Javascipt or Go. The down side is you might need to define convert methods or data types for some exports to be compatible.

METADATA could then be essentially a staging area for APIs so things could migrate to the standard library when they mature. JuliaGeometry has a very messy set of APIs, but as we start to converge we could register a Geometry.jl package which we could then submit to the standard library as a PR.

**Comment by Rory-Finnegan** [comment]
I like the idea of a curated package index, however, I might suggest allowing something like [Gentoo overlays](https://wiki.gentoo.org/wiki/Overlay) for none curated packages. This way if folks are experimenting with different packages they can just work within an overlay. You could probably add it with something like `Pkg.overlay(<url-of-index>)`.

**Comment by StefanKarpinski** [comment]
@sjkelly, I've been thinking about something like that. The main issue is that you don't really want it on any using statement anywhere since then they can conflict in various parts of your code. What you really want is a single file that says what versions of things to use. To complement that, you want a really easy way to get those versions, at least for that particular project. One thing I've talked a bit with @wildart and @JeffBezanson about is the idea of loading code directly from bare git repos using `libgit2`, but I think that could be a separate thing from this.

**Comment by rened** [comment]
While I for my part am quite happy with having a single `METADATA`, the reason why I created `DeclarativePackages.jl` is close to my heart: non-global, declarative sets of `REQUIRE`s, which are fixed by SHAs and can be checked in into version control.

By providing a `DECLARE` file (a spiced up `REQUIRE`), `DeclarativePackages.jl` on startup makes sure that only and exactly those packages/-versions are available. This greatly helps with team collaboration, debugging, project versioning (older commits are actually usable), stable vs dev, working on different projects (with different version requirements) in parallel etc.

It also allows to use arbitrary git repos, thus having less need for OpenPackages vs Curated and making it easy to use non-public packages.

Ideally, I'd love to have the built-in package management work like that as well (but I am biased of course ;-), with a global, fall-back `REQUIRE` providing the current behavior in case there are no local declarations.

One thing that I always wanted to add was for `DECLARE` to also record the Julia version (0.x at least), and start the corresponding one on start-up, or at least warn the user. Use case: using 1 year old code without going insane.

**Comment by sbromberger** [comment]
As a relative newcomer, I am not in favor of having an open registry. One of the huge benefits of the current approach is that newbies like myself are gently corrected when we try to do something wrong. Here are a few examples from my own work:

- @StefanKarpinski suggesting that "v1.0.0" was not an appropriate version number for a new package;
- @mlubin suggesting a different name for my graphs package (was originally `FastGraphs.jl`;
- @IainNZ and others reminding me that I needed a license file;
- Many others screaming when my package merge request resulted in 57+ changes to other packages

In short, a free-for-all sounds great for folks who have experience with the existing closely-curated approach, but would be horrible without that previous understanding of "the way things should be done".

**Comment by milktrader** [comment]
Yeah, I the more I think about it the more we should let someone *else* create an OpenRegistry styled repo of repos, sort of like RubyGems does for Ruby. 

Thanks for the positive feedback @sbromberger !

**Comment by wildart** [comment]
What about adding multiple METADATA repositories to the package management? It could be curated content or a private distribution base.

**Comment by quinnj** [comment]
@wildart, one can already start a custom METADATA to curate a private or otherwise collection. See an old post [here](https://groups.google.com/forum/#!searchin/julia-users/METADATA/julia-users/Mqjhe1gxj44/rVUQb-BVoBgJ).

Really, I've started feeling stronger and stronger that METADATA.jl should be a rigorously curated, 100%-tests-passing collection of packages, and other "unofficial" collections can be started by 3rd parties as they see fit.

**Comment by wildart** [comment]
@quinnj I meant simultaneous use of multiple METADATAs. In this case, we could keep curated content separately from open-for-everyone registry.

**Comment by quinnj** [comment]
I know; I'm saying we already allow that (see the link above), and that anyone could go right now and start a new "Free-For-All-METADATA" that could start accepting entries. What I'm arguing is to leave that decoupled from METADATA.jl and let anyone who feels the need for an open registry system to go start it.

**Comment by wildart** [comment]
Exactly my thought, create any repository with METADATA.jl structure and keep your content there. We just need to provide support (dependency and name clash resolution) for handling a collection of METADATA-like package repositories, not just METADATA.jl branches.

**Comment by quinnj** [comment]
True, true, I do believe `Pkg` needs a little more work to play nicely with custom METADATA repos. We would certainly want to hook into the nice dependency resolution functionality for custom repos.

**Comment by svaksha** [comment]
@quinnj's suggestion of being able to use the `Pkg.add("quinnj/CSV.jl")` command is quite useful for those wanting to install the current git master on their unstable julia version <-> assuming the "to-be-installed" package is also being maintained and tested on julia unstable. I ran into this issue with a dependency of a dependency that kept downgrading (for lack of a better term) the `DataStructures` package version which broke code, until I changed the master `REQUIRE` file.

Hence, it would be really nice if this were possible via the master `REQUIRE` file in the `.julia` directory - atm, it can be pinned to a particular version released by the package owner, but if the package is not tagged with a release version (or for non-registered package that does not conform to METADATA standards, aka, `OpenRegistry` packages) it will be easier to try them out and/or uninstall them.

WRT, `OpenRegistry`, I'll echo those who'd like to concentrate on making `METADATA.jl/CuratedPackages.jl` more user friendly with respect to (un)installing packages.

**Comment by hayd** [comment]
I wonder if it's worth creating a JuliaMeta org at the same time as moving to CuratedPackages... Purely to stop CI builds overlapping/tripping over julia. (It's my understanding that travis build queues are shared over a user/organisation rather than per project.)

**Comment by ggggggggg** [comment]
+1 to allowing non-curated packages to depend on a url.

**Comment by milktrader** [comment]
>This requires a small change to the Pkg.add() method to include a kwarg for repo="CuratedPackages"

Probably worth a PR

**Comment by Nosferican** [comment]
This is moot and can be closed.

**Comment by DilumAluthge** [comment]
METADATA.jl is no longer maintained, so I am closing this issue.

If this issue is still relevant, please re-open a new issue on the Julia General registry: https://github.com/JuliaRegistries/General
